[FLINK-10591][table] Introduced functions to return time attributes from MATCH_RECOGNIZE

dawidwys opened the pull request:
## What is the purpose of the change

This PR introduces sql functions that allow extracting time attributes from `MATCH_RECOGNIZE` clause. It introduces `MATCH_ROWTIME` and `MATCH_PROCTIME` functions. It also switches to newly introduced in cep library `PatternProcessFunction`. This function provides more fine grained access to contextual information and provides rich interface as well.

## Brief change log
- 3483ef0 - static methods `generateIterativeCondition` and `generateOneRowPerMatchExpression` shared a lot parameters and always worked on top of `MatchCodeGenerator` therefore I converted them into instance methods
- 9155542 - switches from `PatternSelectFunction` to `PatternProcessFunction`
- c5cef52e - adds support for `MATCH_ROWTIME` and `MATCH_PROCTIME`
- b64cea1 - describes the functions in documentation
- 83e5f77 - formats match_recognize.md to keep line length under 100 characters. There are just a few places where this limit is not restricted as that would break the markdown rendering (long markdown links or html tags)


## Verifying this change

It adds tests in `MatchRecognizeITCase`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **yes**
  - If yes, how is the feature documented? **docs**

dawidwys commented:
@twalthr I would appreciate your review, whenever you have some spare cycles ;)

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/calcite/RelTimeIndicatorConverter.scala:
```diff
@@ -490,13 +503,27 @@ class RexTimeIndicatorMaterializer(
+        operand match {
+          case operand: RexCall if operand.getOperator == BasicOperatorTable.MATCH_PROCTIME ||
+            operand.getOperator == BasicOperatorTable.MATCH_ROWTIME =>
+            operand
```
Why do we return `operand` here?

pnowojski commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/sql/MatchRecognizeITCase.scala:
```diff
@@ -63,7 +64,7 @@ class MatchRecognizeITCase extends StreamingWithStateTestBase {
 
     val sqlQuery =
       s"""
-        |SELECT T.aid, T.bid, T.cid
```
is this change related to the rest of the commit?

pnowojski commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/sql/MatchRecognizeITCase.scala:
```diff
@@ -311,6 +312,61 @@ class MatchRecognizeITCase extends StreamingWithStateTestBase {
+    data.+=(("ACME", Time.seconds(5).toMilliseconds, 1, 3))
+
+
+    val t = env.fromCollection(data)
```
please, do not use one letter variables. Can you name `t` based on it's content? Like `sales`? `orders`? `prices`?

pnowojski commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/sql/MatchRecognizeITCase.scala:
```diff
@@ -311,6 +312,61 @@ class MatchRecognizeITCase extends StreamingWithStateTestBase {
+         |  MEASURES
+         |    A.price as price,
+         |    A.tax as tax,
+         |    MATCH_ROWTIME() as tstamp
```
`tstamp` -> `timeStamp`/`rowTime`

pnowojski commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/sql/MatchRecognizeITCase.scala:
```diff
@@ -311,6 +312,61 @@ class MatchRecognizeITCase extends StreamingWithStateTestBase {
+
+
+    val t = env.fromCollection(data)
+      .assignAscendingTimestamps(e => e._2)
```
`e` -> `event`? `saleEvent`? `orderEvent`?

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/calcite/RelTimeIndicatorConverter.scala:
```diff
@@ -25,6 +25,7 @@ import org.apache.calcite.rel.{RelNode, RelShuttle}
 import org.apache.calcite.rex._
 import org.apache.calcite.sql.`type`.SqlTypeName
 import org.apache.calcite.sql.fun.SqlStdOperatorTable
+import org.apache.calcite.sql.fun.SqlStdOperatorTable.FINAL
```
I think changes in this class should also be covered by a smaller unit test asserting the plan, not only by the `ITCase`.

pnowojski commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/sql/MatchRecognizeITCase.scala:
```diff
@@ -647,29 +704,25 @@ class MatchRecognizeITCase extends StreamingWithStateTestBase {
 
-    val expected = List("1")
+    // note sql works on timestamps in gmt
+    val expected = List(ZonedDateTime.now(ZoneId.of("GMT")).getHour.toString)
```
Will this fail in case of current hour advances between `env.execute()` line and this line? I think this test should be moved to unit test form using `TestHarness`, where you can freeze proc time?

pnowojski commented on docs/dev/table/streaming/match_recognize.md:
```diff
@@ -902,6 +902,37 @@ One has to keep in mind that in case of the `SKIP TO FIRST/LAST variable` strate
+Time attributes
+---------------
+
+In order to apply some subsequent queries on top of the `MATCH_RECOGNIZE` it might be required to use [time attributes](time_attributes.html). To select those there
```
Can you split the longer lines (maybe even in the whole file in a separate commit, or at least the newly added)? Longer lines are causing problems with resolving conflicts and make diffs less readable.

dawidwys commented:
Hi @pnowojski,
Thx for the review, I addressed your comments.

As for formatting the match_recognize.md, I formatted most of it to keep line length under 100 characters. There are just a few places where this limit is not restricted as that would break the markdown rendering (long markdown links or html tags).

Would appreciate final/another look ;)

pnowojski commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/harness/MatchHarnessTest.scala:
```diff
@@ -0,0 +1,97 @@
+
+    val expectedOutput = new ConcurrentLinkedQueue[Object]()
+
+    val now = Instant.now()
```
I'm a bit concerned that this test might not be fully reproducible/it is not fully deterministic - for example it can fail on travis because of some special value of `now` that locally was not the same. In this case this is rather a minor issue, but generally speaking it's better if tests are fully deterministic, or that at least you can reproduce them in the exactly same way locally (for example by using random seed that is being printed at the beginning of the test).

pnowojski commented:
Thanks @dawidwys! Merging
